File: Q-tee(과외구함)/backend/services/notification-service/app/services/notification_service.py
from typing import Dict, Any, List
from .redis_client import redis_client
import logging
from datetime import datetime, timezone

logger = logging.getLogger(__name__)

class NotificationService:

    def _create_notification(self, notif_type: str, notif_id: str, data: Dict[str, Any]) -> Dict[str, Any]:
        """공통 알림 객체 생성"""
        return {
            "type": notif_type,
            "id": notif_id,
            "data": data,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "read": False
        }

    def _send_notification(self, receiver_type: str, receiver_id: int, notification: Dict[str, Any]) -> bool:
        """알림 전송 공통 로직"""
        try:
            publish_success = redis_client.publish_notification(receiver_type, receiver_id, notification)
            store_success = redis_client.store_notification(receiver_type, receiver_id, notification)

            if publish_success or store_success:
                logger.info(f"{notification['type']} notification sent to {receiver_type}:{receiver_id}")
                return True
            else:
                logger.warning(f"Failed to send notification to {receiver_type}:{receiver_id}")
                return False
        except Exception as e:
            logger.error(f"Failed to send notification: {e}")
            return False

    def send_message_notification(self, message_data: Dict[str, Any]) -> bool:
        """쪽지 알림 전송"""
        notification = self._create_notification(
            "message",
            f"msg_{message_data['message_id']}_{int(datetime.now().timestamp())}",
            {
                "message_id": message_data['message_id'],
                "sender_id": message_data['sender_id'],
                "sender_name": message_data['sender_name'],
                "sender_type": message_data['sender_type'],
                "subject": message_data['subject'],
                "preview": message_data.get('preview', ''),
                "classroom_id": message_data.get('classroom_id')
            }
        )
        return self._send_notification(message_data['receiver_type'], message_data['receiver_id'], notification)

    def send_problem_generation_notification(self, data: Dict[str, Any]) -> bool:
        """문제 생성 완료 알림"""
        notif_type = "problem_generation_failed" if not data.get('success', True) else "problem_generation"
        notification = self._create_notification(
            notif_type,
            f"prob_gen_{data['task_id']}_{int(datetime.now().timestamp())}",
            {
                "task_id": data['task_id'],
                "subject": data['subject'],
                "worksheet_id": data['worksheet_id'],
                "worksheet_title": data['worksheet_title'],
                "problem_count": data['problem_count'],
                "success": data.get('success', True),
                "error_message": data.get('error_message')
            }
        )
        return self._send_notification(data['receiver_type'], data['receiver_id'], notification)

    def send_problem_regeneration_notification(self, data: Dict[str, Any]) -> bool:
        """문제 재생성 완료 알림"""
        notif_type = "problem_regeneration_failed" if not data.get('success', True) else "problem_regeneration"
        notification = self._create_notification(
            notif_type,
            f"prob_regen_{data['task_id']}_{int(datetime.now().timestamp())}",
            {
                "task_id": data['task_id'],
                "subject": data['subject'],
                "worksheet_id": data['worksheet_id'],
                "worksheet_title": data['worksheet_title'],
                "problem_number": data['problem_number'],
                "success": data.get('success', True),
                "error_message": data.get('error_message')
            }
        )
        return self._send_notification(data['receiver_type'], data['receiver_id'], notification)

    def send_assignment_submitted_notification(self, data: Dict[str, Any]) -> bool:
        """과제 제출 알림 (선생님이 받음)"""
        notification = self._create_notification(
            "assignment_submitted",
            f"assign_submit_{data['assignment_id']}_{data['student_id']}_{int(datetime.now().timestamp())}",
            {
                "assignment_id": data['assignment_id'],
                "assignment_title": data['assignment_title'],
                "student_id": data['student_id'],
                "student_name": data['student_name'],
                "class_id": data['class_id'],
                "class_name": data['class_name'],
                "submitted_at": data['submitted_at']
            }
        )
        return self._send_notification(data['receiver_type'], data['receiver_id'], notification)

    def send_assignment_deployed_notification(self, data: Dict[str, Any]) -> bool:
        """과제 배포 알림 (학생이 받음)"""
        notification = self._create_notification(
            "assignment_deployed",
            f"assign_deploy_{data['assignment_id']}_{int(datetime.now().timestamp())}",
            {
                "assignment_id": data['assignment_id'],
                "assignment_title": data['assignment_title'],
                "class_id": data['class_id'],
                "class_name": data['class_name'],
                "due_date": data.get('due_date')
            }
        )
        return self._send_notification(data['receiver_type'], data['receiver_id'], notification)

    def send_class_join_request_notification(self, data: Dict[str, Any]) -> bool:
        """클래스 가입 요청 알림 (선생님이 받음)"""
        notification = self._create_notification(
            "class_join_request",
            f"class_join_{data['class_id']}_{data['student_id']}_{int(datetime.now().timestamp())}",
            {
                "student_id": data['student_id'],
                "student_name": data['student_name'],
                "class_id": data['class_id'],
                "class_name": data['class_name'],
                "message": data.get('message')
            }
        )
        return self._send_notification(data['receiver_type'], data['receiver_id'], notification)

    def send_class_approved_notification(self, data: Dict[str, Any]) -> bool:
        """클래스 승인 알림 (학생이 받음)"""
        notification = self._create_notification(
            "class_approved",
            f"class_approved_{data['class_id']}_{int(datetime.now().timestamp())}",
            {
                "class_id": data['class_id'],
                "class_name": data['class_name'],
                "teacher_name": data['teacher_name']
            }
        )
        return self._send_notification(data['receiver_type'], data['receiver_id'], notification)

    def send_grading_updated_notification(self, data: Dict[str, Any]) -> bool:
        """채점 수정 알림 (학생이 받음)"""
        notification = self._create_notification(
            "grading_updated",
            f"grading_{data['assignment_id']}_{int(datetime.now().timestamp())}",
            {
                "assignment_id": data['assignment_id'],
                "assignment_title": data['assignment_title'],
                "score": data['score'],
                "feedback": data.get('feedback')
            }
        )
        return self._send_notification(data['receiver_type'], data['receiver_id'], notification)

    def send_market_sale_notification(self, data: Dict[str, Any]) -> bool:
        """마켓 판매 알림 (판매자가 받음)"""
        notification = self._create_notification(
            "market_sale",
            f"market_sale_{data['product_id']}_{int(datetime.now().timestamp())}",
            {
                "product_id": data['product_id'],
                "product_title": data['product_title'],
                "buyer_id": data['buyer_id'],
                "buyer_name": data['buyer_name'],
                "amount": data['amount']
            }
        )
        return self._send_notification(data['receiver_type'], data['receiver_id'], notification)

    def send_market_new_product_notification(self, data: Dict[str, Any]) -> bool:
        """마켓 신상품 알림"""
        notification = self._create_notification(
            "market_new_product",
            f"market_new_{data['product_id']}_{int(datetime.now().timestamp())}",
            {
                "product_id": data['product_id'],
                "product_title": data['product_title'],
                "seller_name": data['seller_name']
            }
        )
        return self._send_notification(data['receiver_type'], data['receiver_id'], notification)

    def send_bulk_notifications(self, notifications_data: List[Dict[str, Any]]) -> int:
        """대량 알림 전송"""
        success_count = 0
        for notif_data in notifications_data:
            # type 필드로 어떤 알림인지 판단
            notif_type = notif_data.get('type', 'message')

            method_map = {
                'message': self.send_message_notification,
                'problem_generation': self.send_problem_generation_notification,
                'problem_regeneration': self.send_problem_regeneration_notification,
                'assignment_submitted': self.send_assignment_submitted_notification,
                'assignment_deployed': self.send_assignment_deployed_notification,
                'class_join_request': self.send_class_join_request_notification,
                'class_approved': self.send_class_approved_notification,
                'grading_updated': self.send_grading_updated_notification,
                'market_sale': self.send_market_sale_notification,
                'market_new_product': self.send_market_new_product_notification
            }

            method = method_map.get(notif_type, self.send_message_notification)
            if method(notif_data):
                success_count += 1

        logger.info(f"Sent {success_count}/{len(notifications_data)} notifications")
        return success_count

    def get_stored_notifications(self, user_type: str, user_id: int, limit: int = 10) -> List[Dict[str, Any]]:
        """저장된 알림 조회"""
        try:
            notifications = redis_client.get_stored_notifications(user_type, user_id, limit)
            logger.info(f"Retrieved {len(notifications)} stored notifications for {user_type}:{user_id}")
            return notifications
        except Exception as e:
            logger.error(f"Failed to get stored notifications: {e}")
            return []

notification_service = NotificationService()

File: Q-tee(과외구함)/Q-tee-Frontend/src/services/notificationService.ts
'use client';

// 알림 타입 정의
export type NotificationType =
  | 'message'
  | 'problem_generation'
  | 'problem_regeneration'
  | 'problem_generation_failed'
  | 'problem_regeneration_failed'
  | 'assignment_submitted'
  | 'assignment_deployed'
  | 'class_join_request'
  | 'class_approved'
  | 'grading_updated'
  | 'market_sale'
  | 'market_new_product';

// 각 알림 타입별 데이터 인터페이스
export interface MessageNotificationData {
  message_id: number;
  sender_id: number;
  sender_name: string;
  sender_type: 'teacher' | 'student';
  subject: string;
  preview: string;
  classroom_id?: number;
}

export interface ProblemGenerationData {
  task_id: string;
  subject: 'math' | 'korean' | 'english';
  worksheet_id: number;
  worksheet_title: string;
  problem_count: number;
  success: boolean;
  error_message?: string;
}

export interface ProblemRegenerationData {
  task_id: string;
  subject: 'math' | 'korean' | 'english';
  worksheet_id: number;
  worksheet_title: string;
  problem_number: number;
  success: boolean;
  error_message?: string;
}

export interface AssignmentSubmittedData {
  assignment_id: number;
  assignment_title: string;
  student_id: number;
  student_name: string;
  class_id: number;
  class_name: string;
  submitted_at: string;
}

export interface AssignmentDeployedData {
  assignment_id: number;
  assignment_title: string;
  class_id: number;
  class_name: string;
  due_date?: string;
}

export interface ClassJoinRequestData {
  student_id: number;
  student_name: string;
  class_id: number;
  class_name: string;
  message?: string;
}

export interface ClassApprovedData {
  class_id: number;
  class_name: string;
  teacher_name: string;
}

export interface GradingUpdatedData {
  assignment_id: number;
  assignment_title: string;
  score: number;
  feedback?: string;
}

export interface MarketSaleData {
  product_id: number;
  product_title: string;
  buyer_id: number;
  buyer_name: string;
  amount: number;
}

export interface MarketNewProductData {
  product_id: number;
  product_title: string;
  seller_name: string;
}

// Union type for all notification data
export type NotificationData =
  | MessageNotificationData
  | ProblemGenerationData
  | ProblemRegenerationData
  | AssignmentSubmittedData
  | AssignmentDeployedData
  | ClassJoinRequestData
  | ClassApprovedData
  | GradingUpdatedData
  | MarketSaleData
  | MarketNewProductData;

export interface SSENotification {
  type: NotificationType;
  id: string;
  data: NotificationData;
  timestamp: string;
  read: boolean;
}

export interface StoredNotificationResponse {
  notifications: SSENotification[];
}

class NotificationService {
  private eventSource: EventSource | null = null;
  private listeners: ((notification: SSENotification) => void)[] = [];
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 1000;
  private baseUrl = process.env.NEXT_PUBLIC_NOTIFICATION_SERVICE_URL || 'http://localhost:8006';

  connect(userType: 'teacher' | 'student', userId: number): void {
    if (this.eventSource) {
      this.disconnect();
    }

    // Uncomment the following code when notification service is implemented:
    
    const url = `${this.baseUrl}/api/notifications/stream/${userType}/${userId}`;

    this.eventSource = new EventSource(url);

    this.eventSource.onopen = () => {
      this.reconnectAttempts = 0;
    };

    this.eventSource.onmessage = (event) => {
      try {
        const notification: SSENotification = JSON.parse(event.data);
        this.notifyListeners(notification);
      } catch (error) {
        // 알림 파싱 에러는 조용히 처리
      }
    };

    this.eventSource.onerror = (error) => {
      this.handleReconnect(userType, userId);
    };
    
  }

  disconnect(): void {
    if (this.eventSource) {
      this.eventSource.close();
      this.eventSource = null;
    }
    this.reconnectAttempts = 0;
  }

  private handleReconnect(userType: 'teacher' | 'student', userId: number): void {
    if (this.reconnectAttempts < this.maxReconnectAttempts) {
      this.reconnectAttempts++;
      const delay = this.reconnectDelay * Math.pow(2, this.reconnectAttempts - 1);

      setTimeout(() => {
        this.connect(userType, userId);
      }, delay);
    }
  }

  addListener(callback: (notification: SSENotification) => void): void {
    this.listeners.push(callback);
  }

  removeListener(callback: (notification: SSENotification) => void): void {
    this.listeners = this.listeners.filter(listener => listener !== callback);
  }

  private notifyListeners(notification: SSENotification): void {
    this.listeners.forEach(listener => listener(notification));
  }

  async getStoredNotifications(
    userType: 'teacher' | 'student',
    userId: number,
    limit: number = 10
  ): Promise<SSENotification[]> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/stored/${userType}/${userId}?limit=${limit}`
      );

      if (!response.ok) {
        throw new Error(`HTTP error! status: ${response.status}`);
      }

      const data: StoredNotificationResponse = await response.json();
      return data.notifications;
    } catch (error) {
      return [];
    }
  }

  async clearStoredNotifications(userType: 'teacher' | 'student', userId: number): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/stored/${userType}/${userId}`,
        { method: 'DELETE' }
      );

      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async markAsRead(userType: 'teacher' | 'student', userId: number, notificationId: string): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/stored/${userType}/${userId}/${notificationId}/read`,
        { method: 'PATCH' }
      );
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async markAllAsRead(userType: 'teacher' | 'student', userId: number): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/stored/${userType}/${userId}/read-all`,
        { method: 'PATCH' }
      );
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async deleteNotificationsByType(userType: 'teacher' | 'student', userId: number, notificationType: string): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/stored/${userType}/${userId}/type/${notificationType}`,
        { method: 'DELETE' }
      );
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async deleteNotification(userType: 'teacher' | 'student', userId: number, notificationType: string, notificationId: string): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/stored/${userType}/${userId}/${notificationType}/${notificationId}`,
        { method: 'DELETE' }
      );
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async sendTestNotification(userType: 'teacher' | 'student', userId: number): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/test/${userType}/${userId}`
      );

      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async sendProblemGenerationNotification(
    data: ProblemGenerationData & { receiver_id: number; receiver_type: 'teacher' | 'student' }
  ): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/problem/generation`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(data)
        }
      );
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async sendProblemRegenerationNotification(
    data: ProblemRegenerationData & { receiver_id: number; receiver_type: 'teacher' | 'student' }
  ): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/problem/regeneration`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(data)
        }
      );
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async sendAssignmentSubmittedNotification(
    data: AssignmentSubmittedData & { receiver_id: number; receiver_type: 'teacher' }
  ): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/assignment/submitted`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(data)
        }
      );
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async sendAssignmentDeployedNotification(
    data: AssignmentDeployedData & { receiver_id: number; receiver_type: 'student' }
  ): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/assignment/deployed`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(data)
        }
      );
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async sendClassJoinRequestNotification(
    data: ClassJoinRequestData & { receiver_id: number; receiver_type: 'teacher' }
  ): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/class/join-request`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(data)
        }
      );
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async sendClassApprovedNotification(
    data: ClassApprovedData & { receiver_id: number; receiver_type: 'student' }
  ): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/class/approved`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(data)
        }
      );
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async sendGradingUpdatedNotification(
    data: GradingUpdatedData & { receiver_id: number; receiver_type: 'student' }
  ): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/grading/updated`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(data)
        }
      );
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async sendMarketSaleNotification(
    data: MarketSaleData & { receiver_id: number; receiver_type: 'teacher' | 'student' }
  ): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/market/sale`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(data)
        }
      );
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  async sendMarketNewProductNotification(
    data: MarketNewProductData & { receiver_id: number; receiver_type: 'teacher' | 'student' }
  ): Promise<boolean> {
    try {
      const response = await fetch(
        `${this.baseUrl}/api/notifications/market/new-product`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(data)
        }
      );
      return response.ok;
    } catch (error) {
      return false;
    }
  }

  isConnected(): boolean {
    // TODO: Return false until notification service is implemented
    return false;
    // return this.eventSource !== null && this.eventSource.readyState === EventSource.OPEN;
  }
}

export const notificationService = new NotificationService();